Improve the JSDoc comments

timvandermeij opened the pull request:
This commit series makes our JSDoc comments more consistent and allows us to generate better API documentation and TypeScript definition files.

I have been watching the discussions about both for some time now and figured we should get started with making improvements because I think it would be ideal from a maintenance perspective if we could generate them both instead of maintaining them manually. Therefore, I collected  improvements both from the various discussions and the JSDoc documentation so the resulting output becomes more precise.

Note that this is groundwork for more documentation improvements, so more work definitely needs to be done. Thanks to the various people that discussed this and did proposals for this.

Before:
https://mozilla.github.io/pdf.js/api/draft/PDFDocumentProxy.html

After:
http://54.67.70.0:8877/15dd7d913e2aab9/api/draft/module-pdfjsLib-PDFDocumentProxy.html

(notice that all getters now have types and descriptions)

timvandermeij committed:
Consistently use `number` for numeric data types in JSDoc comments

Sometimes we also used `Number` and `integer`, but `number` is what
the JSDoc documentation recommends.

timvandermeij committed:
Consistently use `string` for string data types in JSDoc comments

Sometimes we also used `String`, but `string` is the what the JSDoc
documentation recommends.

timvandermeij committed:
Consistently use square brackets for optional parameters in JSDoc comments

Square brackets are recommended to indicate optional parameters. Using
them helps for automatically generating correct documentation.

timvandermeij committed:
Consistently use `@type` for getter data types in JSDoc comments

Sometimes we also used `@return` or `@returns`, but `@type` is what
the JSDoc documentation recommends. This also improves the documentation
because before this commit the types were not shown and now they are.

timvandermeij committed:
Consistently use `@returns` for returned data types in JSDoc comments

Sometimes we also used `@return`, but `@returns` is what the JSDoc
documentation recommends. Even though `@return` works as an alias, it's
good to use the recommended syntax and to be consistent within the
project.

timvandermeij committed:
Bundle all API documentation in a module

This commit allows JSDoc to generate all API documentation in the
`pdfjsLib` module (namespace) so the documentation becomes easier to
navigate.

timvandermeij committed:
Don't include `src/shared/util.js` for the API documentation

This file only contains helper functions and should not be listed in the
documentation since they are not part of the public API.

timvandermeij commented:
/botio-linux preview

Snuffleupagus commented on src/display/api.js:
```diff
@@ -15,6 +15,10 @@
 /* eslint no-var: error */
 
+/**
+ * @module pdfjs
```
Given that the `PDFJS` global was removed, and that we're now using `pdfjsLib` for code running on the main-thread, is that name really appropriate at this point!?

I readily admit not knowing much about JSDoc `@module`, but this was something that stood out to me.

timvandermeij replied:
I checked how we import `pdfjs-dist` in the examples and we indeed use `pdfjsLib` there, so that sounds like a better name for the documentation too. I have changed this.

timvandermeij commented:
/botio-linux preview

Snuffleupagus commented:
Another thing I just noticed is that `PDFWorker` is now listed *twice* under the "Classes" heading.

timvandermeij commented:
Yes, this is because the closure is called `PDFWorker` and the inner class is called `PDFWorker`. If you click the links, they are in fact different. Looking at this again, maybe we should leave out `PDFWorker` entirely, since normally API users shouldn't need it?

Snuffleupagus commented:
> Looking at this again, maybe we should leave out `PDFWorker` entirely since I don't think API consumers should need it, or do you think there is a use case for including it?

Given https://github.com/mozilla/pdf.js/blob/4af2755f714f562787ed4302dbc5c599a4056f39/src/display/api.js#L146-L147 you probably need to keep it. However, couldn't you fix things simply by moving the docs to https://github.com/mozilla/pdf.js/blob/4af2755f714f562787ed4302dbc5c599a4056f39/src/display/api.js#L1593-L1596

timvandermeij committed:
Deduplicate the documentation for `PDFDocumentLoadingTask` and `PDFWorker`

Both classes live inside a closure with the same name, which confuses
JSDoc. Move the documentation to the inner class to deduplicate them.

timvandermeij committed:
Don't include private methods in the the `PDFPageProxy` API documentation

timvandermeij commented:
/botio-linux preview

Snuffleupagus commented on src/display/api.js:
```diff
@@ -1590,9 +1583,15 @@ const PDFWorker = (function PDFWorkerClosure() {
 
   /**
-   * @param {PDFWorkerParameters} params - The worker initialization parameters.
+   * The web worker abstraction controls the instantiation of PDF documents.
```
Please keep the `PDF.js`-part of this line, perhaps as `PDF.js web worker abstraction, which controls the instantiation of PDF documents.`, since this is different from the general concept of the [Web Workers API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API).

timvandermeij replied:
Done. Thanks!

Snuffleupagus reviewed (approved):
Looks fine by me, modulo one small last comment; thanks!
